src/wavm/io.cpp

```diff
@@ -338,23 +339,39 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(wasi,
                                I32 iovecCount,
                                I32 resBytesWrittenPtr)
 {
-    faabric::util::getLogger()->debug("S - fd_write - {} {} {} {}",
+    storage::FileSystem& fileSystem = getExecutingWAVMModule()->getFileSystem();
+    std::string path = fileSystem.getPathForFd(fd);
+
+    faabric::util::getLogger()->debug("S - fd_write - {} {} {} {} ({})",
                                       fd,
                                       iovecsPtr,
                                       iovecCount,
-                                      resBytesWrittenPtr);
+                                      resBytesWrittenPtr,
+                                      path);
 
-    storage::FileDescriptor& fileDesc =
-      getExecutingWAVMModule()->getFileSystem().getFileDescriptor(fd);
+    storage::FileDescriptor& fileDesc = fileSystem.getFileDescriptor(fd);
 
     iovec* nativeIovecs = wasiIovecsToNativeIovecs(iovecsPtr, iovecCount);
+
     ssize_t bytesWritten =
-      doWritev(fileDesc.getLinuxFd(), nativeIovecs, iovecCount);
+      ::writev(fileDesc.getLinuxFd(), nativeIovecs, iovecCount);
 
-    Runtime::memoryRef<int>(getExecutingWAVMModule()->defaultMemory,
-                            resBytesWrittenPtr) = (int)bytesWritten;
+    // Catpure stdout if necessary, otherwise write as normal
+    faabric::util::SystemConfig& conf = faabric::util::getSystemConfig();
+    if (fd == STDOUT_FILENO && conf.captureStdout == "on") {
+        getExecutingWAVMModule()->captureStdout(nativeIovecs, iovecCount);
+    }
 
-    return __WASI_ESUCCESS;
+    Runtime::memoryRef<uint64_t>(getExecutingWAVMModule()->defaultMemory,
+                                 resBytesWrittenPtr) = bytesWritten;
+
+    delete[] nativeIovecs;
+
+    if (bytesWritten < 0) {
+        return storage::errnoToWasi(errno);
+    } else {
+        return __WASI_ESUCCESS;
```

Comment:
This function was ignoring errors, and not correctly setting the number of bytes written.